Servers cache fix

SparrowBrain opened the pull request:
### Issue
Currently the plugin stores all websocket servers it receives from Steam in a cache. During the startup if connection to the first server from cache fails the plugin will try the next one in the saved list. If the connection fails again, the plugin will keep iterating until it succeeds or we reach authentication timeout.

Sometimes the cached Steam servers become unreachable. This can lead to an issue, where we are iterating through a list, but don't get a working websocket server before the timeout.

Because of this, plugin stays offline. The only way to refresh cache is to wait for 30 days after it was added.

### Implementation
The changes consist of:
* We're only saving one websocket server per `cell_id`. This is done because if the first server in the cache becomes invalid, we should consider the whole cache of servers invalid (untrustworthy). Iterating through it becomes guesswork, since we don't know how many more of the servers are invalid. Best thing to do is to refresh the list and cache newest data.
* `servers_cache` now becomes `websocket_cache`. This was done to remove potential data formatting issues and to make it easier to write code in TDD fashion.
* Added a `#TODO` to remove `servers_cache` cleanup code after some 6 months.
* `websocket_cache` splits the responsibilities of `servers_cache` into 3 separate components.

#### `websocket_cache_persistence` 
Reposponsible for formatting and writting/reading to/from cache

#### `websocket_cache_list` 
Responsible for fetching servers from Steam and ordering them by lowest ping

#### `websocket_cache` 
The coordinator component that glues everything together. It first `yield` returns the cached server if available, then goes on to fetch the server list from Steam and `yield` the results.

`websocket_cache` is populated when returning these newly fetched websockets. Ideally we would do that after we connect with `websocket_client`, but since we've already connected to these sockets to measure the ping, I think this will work most of the time. It also keeps the complexity out of `websocket_client`.


### Considerations
We could keep `servers_cache` name since `websocket_cache_persistence` handles any exceptions during reading.


### Testing
To test either apply the changes manually, or download [servers_cache_fix.zip](https://github.com/FriendsOfGalaxy/galaxy-integration-steam/files/5885996/servers_cache_fix.zip) (based on v0.55) and replace plugin contents in `%LOCALAPPDATA%\GOG.com\Galaxy\plugins\installed\steam_ca27391f-2675-49b1-92c0-896d43afa4f8`

SparrowBrain commented:
Special version for @ReiTung 😄 
[servers_cache_fix_with_invalid_credentials.zip](https://github.com/FriendsOfGalaxy/galaxy-integration-steam/files/5887381/servers_cache_fix_with_invalid_credentials.zip)

This is basically v0.55 + https://github.com/FriendsOfGalaxy/galaxy-integration-steam/pull/103 + this change

ReiTung commented:
❤️❤️❤️

FriendsOfGalaxy commented:
Hi @SparrowBrain 

Its really nice to see well structured and described change.

I have some doubts for business logic (about if we really benefit with your change much over disabling cache at all), but maybe give let me do proper review in next week and do not say anything specific yet.

SparrowBrain commented:
It's ok, I'm all for the 'less is more' mantra - less code we have the better ;).

I'll wait for now with cell_id fix. It's pretty straight forward, we just need to agree where it would be stored, and it might depend on the outcome of this.

FriendsOfGalaxy commented:
Ok, so I've tested this once with having connection problems and has observed interesting things.

First of all thanks to you sorting by ping and logging things, I've seen that for cell_id=0 my pings are like
500ms
520
518
640
670
...

so apparently steam returns already sorted list of servers, probably based on geolocation.

then for cell_id=1, looks like server pings are generally worse, in my case:
900ms
945
1023
...

Based on both info I think `cell_id` is nothing more than iterator though next 100 nearest servers:
- there is no point to increment it on failure
- there is no point of structuring cache around that value

not iterating cell_id should greatly improve current situation. In case 2FA is required, first connection is closed and we reconnect to LogOn again - this time with 2FA code. So most of connection problems may be settled here: cell_id=1 or even bigger is more problematic, and so connecting with steam guard makes experience with this plugin worse.

-------

Second thought is about saving resources by using cache. Testing 100 servers for ping is already quite heavy but we can skip it as they are already - lets say - sorted.

We can save cache in memory only or in persistent cache. The first solution seems more robust for cases like traveling. Imagine you've just move to another country - why not to fetch fresh list of nearest servers?

On the other hand using NetHook2 we can see that steam sends us list of servers `Msg.ClientCMList` quickly after logging. This _may be_ their solution for "traveling" problem. 

But unless we investigate it more, the easiest way to proceed IMO is to:
- re-fetch server list every time on plugin start
- keep in-memory cache to iterate over it in case of any problem

---

Further improvements may be about 
- caching Good and Bad servers like other do (recap my comment here: https://github.com/FriendsOfGalaxy/galaxy-integration-steam/issues/86#issuecomment-704403997)
- researching how to which server Steam client connects after rebooting (do they cache?)
- handle steam errors properly like eresult 63 (2FA, no need to change server) or eresult 48 (TryDiffernetCM) - do not raise error from auth interface, just try next server (I planned bigger refactor here).


I'm curious about your thoughts @SparrowBrain

SparrowBrain commented:
`cell_id` basically represents the download region from Steam.

After the login in `ClientLogOnResponse` we receive `cell_id` Steam thinks we should use. See [this comment](https://github.com/DoctorMcKay/node-steam-user/issues/39#issuecomment-165877056) and the [CellMap](https://github.com/SteamDatabase/SteamTracking/blob/master/ClientExtracted/steam/cached/CellMap.vdf)

I think it would make sense to use that `cell_id` during next login, hence we probably want to store it on disk.
>
>     * there is no point to increment it on failure
Yes, it makes no sense incrementing `cell_id`. 
>     * there is no point of structuring cache around that value
We want to store `cell_id`. `websocket_cache` structure could instead be a single entry:
```
cache = {
            'cell_id': cell_id,
            'server': server,
            'timeout': time.time() + CACHE_ENTRY_TIMEOUT_DAYS * SECONDS_IN_A_DAY
        }
```

I have a version of plugin that stores `cell_id`, but in `UserCacheInfo`. That is probably incorrect, and some `server_cache` would work better.
_____________



> But unless we investigate it more, the easiest way to proceed IMO is to:
> 
>     * re-fetch server list every time on plugin start
> 
>     * keep in-memory cache to iterate over it in case of any problem
I agree, this + remove testing all the servers. This would be the easiest.
___
> Further improvements may be about
> 
>     * caching Good and Bad servers like other do (recap my comment here: [#86 (comment)](https://github.com/FriendsOfGalaxy/galaxy-integration-steam/issues/86#issuecomment-704403997))
I really don't like marking servers as Good or Bad, since just fetching a new list is a lot easier and probably faster. As I've mentioned in PR, I would treat cache as invalid at that point.
>     * researching how to which server Steam client connects after rebooting (do they cache?)
I lack knowledge here honestly. Good thing to do.
>     * handle steam errors properly like eresult 63 (2FA, no need to change server) or eresult 48 (TryDiffernetCM) - do not raise error from auth interface, just try next server (I planned bigger refactor here).
This seems like a good improvement.
____
I guess I would aim for this solution:
* Remove iterating over `cell_id`
* Remove testing all servers during fetching
* Restructure `websocket_cache` to contain `cell_id`. Populate `cell_id` after first logon.
* Keep single server entry in cache to avoid fetching all servers

And for future PR's:
* Handle `Msg.ClientCMList`. If it lists servers Steam want us to use, cache it on disk.
* Improve Steam error handling

@FriendsOfGalaxy I assume this would be rather fast for users and fairly simple on the complexity side. What do you think?

ReiTung commented:
so, first test results with v0.55 special edition (i.e. +#108 +steam lib cache) are in. eresult 5 one day after fresh installation - but also first time with 2 PCs active and connected at the same time (2 different privIPs behind a router with the same extIP). So, not sure if the usual #74 issue or the one coming from multiple machines (believe it was #104). However, I believe error message returned should be different for the multiPC issue, i.e. not eresult 5? Logs attached.

SteamLib Caching after reconnect worked like a charm though!

thanks all,
ReiTung

P.S.: will also post to #74 (same set of logs)

[logs_with_crashed_v053_before_fresh_install_20210122.zip](https://github.com/FriendsOfGalaxy/galaxy-integration-steam/files/5937833/logs_with_crashed_v053_before_fresh_install_20210122.zip)
[logs_with_offline_v055_specialedition_after_fresh_install_pc1_20210206.zip](https://github.com/FriendsOfGalaxy/galaxy-integration-steam/files/5937834/logs_with_offline_v055_specialedition_after_fresh_install_pc1_20210206.zip)
[logs_with_offline_v054_before_fresh_install_20210122.zip](https://github.com/FriendsOfGalaxy/galaxy-integration-steam/files/5937835/logs_with_offline_v054_before_fresh_install_20210122.zip)

FriendsOfGalaxy commented:
> `cell_id` basically represents the download region from Steam.

Thanks, good to know!

> After the login in `ClientLogOnResponse` we receive `cell_id` Steam thinks we should use.

Looks correct based on my location and map you've provided.

> I think it would make sense to use that `cell_id` during next login, hence we probably want to store it on disk.

Looks reasonable as well. But Steam client use different cell_id when sending LogOn, have a look:
![cell_id_from_steam](https://user-images.githubusercontent.com/48480506/107256884-5b756600-6a3a-11eb-96c1-1d3347fbcee1.png)
Maybe this is for more accurate (smaller) region?

Of course it doesn't mean steam used this `cell_id` to fetch servers list via http.

Using your function to test pings, both cell_id's (send via `LogOn`, and received from `LogOnResponse`) returns similar addresses (and pings):

```
wss://cm1-waw1.cm.steampowered.com:27030/cmsocket/ 1363
wss://cm1-waw1.cm.steampowered.com:27039/cmsocket/ 1364
...
```
```
wss://cm1-waw1.cm.steampowered.com:27037/cmsocket/ 1344
wss://cm2-waw1.cm.steampowered.com:27038/cmsocket/ 1344
...
```
Similar for `cell_id=0`.

That is why I prefer using default 0 - looks like autodetection. Maybe asking for concrete cell_id is more accurate. I don't know. But it doesn't look like major optimization for now.

When I use for example cell_id=9 then it lists Taiwan servers with much worse pings.

----

Steam `serverlist_websockets` contains list of server:port
If `wss://cm1-waw1.cm.steampowered.com:27030/cmsocket/` does not respond quickly does it mean wss://cm1-waw1.cm.steampowered.com:27039/cmsocket/ won't respond as well?


> ```
> 
> I have a version of plugin that stores `cell_id`, but in `UserCacheInfo`. That is probably incorrect, and some `server_cache` would work better.

agree. But - keep reading.

> 
> > Further improvements may be about
> > ```
> > * caching Good and Bad servers like other do (recap my comment here: [#86 (comment)](https://github.com/FriendsOfGalaxy/galaxy-integration-steam/issues/86#issuecomment-704403997))
> > ```
> 
> I really don't like marking servers as Good or Bad, since just fetching a new list is a lot easier and probably faster. As I've mentioned in PR, I would treat cache as invalid at that point.

yes, unless fresh list contains "Bad" servers as well. Then you may have a infinite loop - especially if you want to cache only one server. We would need more tests or ask guys from those Steam projects for reasons of such design.

> 
> > ```
> > * researching how to which server Steam client connects after rebooting (do they cache?)
> > ```
> 
> I lack knowledge here honestly. Good thing to do.

Quick search in <Steam installation dir> - there is `cache/` with some interesting things (but mostly looks like consts) and `update_hosts_cached.vdf`  - there is `ClientInfo` with "valid_until" timestamp and ip_country_code. 

We can check when cache files are created/updated.


> 
> > ```
> > * handle steam errors properly like eresult 63 (2FA, no need to change server) or eresult 48 (TryDiffernetCM) - do not raise error from auth interface, just try next server (I planned bigger refactor here).
> > ```
> 
> This seems like a good improvement.

BTW I wonder if TryDifferentCM means different server or just socket (which address contains server and port)

> 
> I guess I would aim for this solution:
> 
> * Remove iterating over `cell_id`
> * Remove testing all servers during fetching

yup. 
There is no need for 3 layers now, I'm not sure if 2 or just 1 is ok. I like your TDD approach so feel free to decide.
BTW It sounds like reverting part of https://github.com/FriendsOfGalaxy/galaxy-integration-steam/commit/9978f477afe39c0d97a57320fb199be7205f076d.

> * Restructure `websocket_cache` to contain `cell_id`. Populate `cell_id` after first logon.

I'm not convinced yet. We should behave like Steam client, but we need to figure out what cell_id should be sent (apparently different `cell_id` than received recently from `LogOnResponse`)

If not, what caching cell_id gives us? We can potencialy detect if user has moved to another region if `cell_id` does not match with cache value - so we can re-fetch server list. But we re-fetch on start anyway.

> * Keep single server entry in cache to avoid fetching all servers

1. You have to fetch all at once anyway - it is just a REST API.
2. possible infinite bad-server loop(?) - mentioned previously


> And for future PR's:
> 
> * Handle `Msg.ClientCMList`. If it lists servers Steam want us to use, cache it on disk.

agree. For now I do not understand its content.

SparrowBrain commented:
Ok, I wasn't aware Steam does something weird with `cell_id`. Thanks for pointing that out.

You have made a lot of good points. So until we understand `cell_id` and the `Msg.ClientCMList` the best thing to do:
* Remove `cell_id` looping. Maybe even some of https://github.com/FriendsOfGalaxy/galaxy-integration-steam/commit/9978f477afe39c0d97a57320fb199be7205f076d logic
* Remove server testing.
* Remove all server caching to disk.

This solves the immediate issue people have with current cache. And leaves more  complicated issues to be researched and solved later.

I'll get on it. Thanks, @FriendsOfGalaxy !

SparrowBrain commented:
Removed server caching and testing of servers. Also removed incrementing `cell_id`.

Elected to keep the actual `cell_id` parameter, since that adds more context to the code.

Any feedback is welcome :).

FriendsOfGalaxy commented on tests/test_websocket_list.py:
```diff
@@ -0,0 +1,33 @@
+
+
+@pytest.mark.asyncio
+async def test_get_ordered_by_ping_returns_single_server(backend_client):
```
no longer ordered by ping

FriendsOfGalaxy replied:
What about a few items to test if the first one was returned instead of one-element list? If you want to test more cases pytest.mark.paramterize is nice way to go.

SparrowBrain replied:
Good catch. Renamed test name, also switched it to test for a list instead of a single entry.

FriendsOfGalaxy commented on tests/test_websocket_list.py:
```diff
@@ -0,0 +1,33 @@
+
+
+@pytest.mark.asyncio
+async def test_get_ordered_by_ping_queries_list_with_given_cell_id(backend_client):
```
no longer ordered by ping

SparrowBrain replied:
Good catch.

FriendsOfGalaxy commented on tests/test_websocket_list.py:
```diff
@@ -0,0 +1,33 @@
+
+    await websocket_list.get(cell_id)
+
+    backend_client.get_servers.assert_called_once_with(cell_id)
```
What about `backend_client.side_effect = BackendError()`? Is it layer above supposed to handle that? (I think so bu t don't remember)

What if that would happen on second `get` in a row? 
Maybe add test case that tests using of the in-memory cache. (`backend_client.get_servers.assert_not_called()`)

SparrowBrain replied:
`BackendError` is handled by layer above (`websocket_client`). I don't think `backend_client.get_servers()` will throw it though. I think all errors that occur in `backend_client.get_servers()` will bubble up right to `plugin`. As it did before.

As for in-memory cache, I'm not sure we understand it the same way. Basically it's just an in-memory variable the `websocket_client_ensure_connected()` iterates through.

The rough structure is:
```
websocket_client
  def run()
    while True:
      try:
        self._ensure_connected():
      except: ....

  def _ensure_connected():
    while True:
    for server in servers
      try:
        connect()
      except: ....
```

So there are two `while True` loops with various exception handling. If the `_ensure_connected()` loop fails, the external loop would cause `get()` to be called again. Which is probably desired.

Also `websocket_client` already has some test for error handling, like `test_servers_cache_retry`.

But overall I don't feel very confident changing large parts of `websocket_client` without some global acceptance tests. And this change is intended to help people with `servers_cache` not working anymore and forcing plugin into 'Offline'.

FriendsOfGalaxy replied:
> As for in-memory cache, I'm not sure we understand it the same way. Basically it's just an in-memory variable the `websocket_client_ensure_connected()` iterates through.

You're right, I've forgotten that there are 2 for loops, hence difference in understanding usage of in-memory cache. To sum up:
- your solution: each time we lost connection (first loop) we fetch fresh list of servers and iterate over them starting from the first one steam backend has returned
- my previous proposal: each time we lost connection, we try to connect with next server in the servers cache (that is populated only once). Technically it could be done by using generator outside `_ensure_conneted`.

My solution is better in hypocritical case where first steam server on the list has worse quality than subsequent servers. Your solution is much better in case user has switched location (in real or VPN) or plugin was running for a longer time.

So let us stick with your solution.

>  I don't think `backend_client.get_servers()` will throw it though

BackendError may be raised by `AuthenticatedHttpClient` by https://github.com/gogcom/galaxy-integrations-python-api/blob/master/src/galaxy/http.py#L113

It is handled perfectly fine, in case of any troubles, we wait `RECONNECT_INTERVAL_SECONDS` and try again.

FriendsOfGalaxy replied:
Just adding comment - we could create some solution between mine and yours, like:
- fetch for new servers list at most once per 10min
- 
but maybe do not optimize prematurely.

FriendsOfGalaxy reviewed (changes requested):
@SparrowBrain I left a few minor comments and request to add test

FriendsOfGalaxy reviewed (approved):
It is cool, merging

ReiTung commented:
@FriendsOfGalaxy what is the correct way to PREVENT an update? just updating the first line in the version file? As long as #74 is still an issue and would trigger complete reimports of Steam data every few days, I want to stay on the "special" version of v0.55 which used the cached data and does not screw up things.
also, from my impression, #74 and the reimport is causing a lot of frustration with people and I recently read quite a few comments stating they turned or are turning away from Galaxy as this is a core connector expected to work ... 
understood and appreciated that the fix is difficult, but if we could at least get rid of the re-import of Steam data that might get some pressure out as reconnecting is quick and people should not care too much about the "red dot" as it is also showing up if other integrations break (even the built-in Epic one btw)

thanks
ReiTung

FriendsOfGalaxy commented:
>  what is the correct way to PREVENT an update? just updating the first line in the version file?

yes, that should work

> also, from my impression, #74 and the reimport is causing a lot of frustration with people and I recently read quite a few comments stating they turned or are turning away from Galaxy as this is a core connector expected to work ...
understood and appreciated that the fix is difficult, but if we could at least get rid of the re-import of Steam data that might get some pressure out as reconnecting is quick and people should not care too much about the "red dot" as it is also showing up if other integrations break (even the built-in Epic one btw)

I understand. I'll back to #74 after next release, probably in the next week.
